tests/cpp/common/test_hist_util.cu

```diff
@@ -146,7 +157,8 @@ void TestCategoricalSketch(size_t n, size_t num_categories, int32_t num_bins) {
 }
 
 TEST(HistUtil, DeviceSketchCategoricalFeatures) {
-  TestCategoricalSketch(1000, 256, 32);
+  TestCategoricalSketch(1000, 256, 32, false);
```

Comment:
How does this test pass, if there are 256 categories and only 32 bins. Then surely `n_uniques` should be 256 and won't be equal to `cuts.TotalBins()`?

Reply:
Bin values for each feature are the unique values.

Reply:
There's no sketching if feature type is categorical.

Reply:
That means `num_bin` is ignored for categorical features.

Reply:
> That means `num_bin` is ignored for categorical features.

Can you point me to the piece of code that detects whether a given feature is categorical? Is it the following snippet?
https://github.com/dmlc/xgboost/blob/b05073bda5a5b0ab6be6281212037e26ed9c0404/src/common/quantile.cu#L626-L635

Reply:
@hcho3  For the GPU implementation of GK sketching, there are 4 parts:

- hist_utils.cu(h).  This file is the front end of GPU sketch, which extracts sliding windows, remove duplicated categorical values, then push the value into quantile structure.
- Pruning.  This is defined in quantile.cu, the implementation is exactly as described in xgboost paper (more exact than CPU implementation).  For categorical data, the first part has already handle everything, so extra code in pruning for cat data, is to ignore it (don't prune cat features).  Search the use of `IsCat` function.
- Merging.  I haven't don't anything there yet.  It's used by data iter, which is used by `DaskDeviceQuantileDMatrix` and `DeviceQuantileDMatrix`.  Also it's required for distributed setting where we need a global sketch for data.
- Creating the bin values.  That's the code you referenced.

Reply:
Hope that helps.

Reply:
Thanks for the detailed reply. Searching for `IsCat` function makes it clear how the sketching algo treats categorical features differently.

> I haven't done anything there yet. It's used by data iter, which is used by DaskDeviceQuantileDMatrix and DeviceQuantileDMatrix.

So fair to say that we can't yet use categorical data in DeviceQuantileDMatrix ?

Reply:
@hcho3 DQD is not yet supported.  See items in https://github.com/dmlc/xgboost/issues/6503 .